core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,420 @@
+            this.added = Collections.unmodifiableList(added);
+            this.removed = Collections.unmodifiableList(removed);
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeList(added);
+            out.writeList(removed);
+        }
+
+        @Override
+        public IndexGraveyard apply(final IndexGraveyard old) {
+            final List<Tombstone> oldQueue = old.tombstones;
+            final Deque<Tombstone> newQueue = new ArrayDeque<>(oldQueue);
+            for (Tombstone tombstone : removed) {
+                newQueue.remove(tombstone);
+            }
+            for (Tombstone tombstone : added) {
+                newQueue.addLast(tombstone);
+            }
+            return new IndexGraveyard.Builder(newQueue).build();
+        }
+
+        /** The index tombstones (deletions) that were added between two states */
+        public List<Tombstone> getAdded() {
+            return added;
+        }
+
+        /** The index tombstones (deletions) that were removed between two states */
+        public List<Tombstone> getRemoved() {
+            return removed;
+        }
+    }
+
+    /**
+     * An individual tombstone entry for representing a deleted index.
+     */
+    final public static class Tombstone implements ToXContent, StreamInputReader<Tombstone>, Writeable<Tombstone> {
+
+        public static final Tombstone PROTOTYPE = new Tombstone(new Index("_na_", "_na_"), 0L);
+        private static final String INDEX_KEY = "index";
+        private static final String DELETE_DATE_KEY = "delete_date";
+        private static final ObjectParser<Tombstone.Builder, Void> TOMBSTONE_PARSER = new ObjectParser<>("tombstoneEntry");
+        static {
+            TOMBSTONE_PARSER.declareObject(Tombstone.Builder::index, Index.INDEX_PARSER, new ParseField(INDEX_KEY));
+            TOMBSTONE_PARSER.declareLong(Tombstone.Builder::deleteDate, new ParseField(DELETE_DATE_KEY));
+        }
+
+        private final Index index;
+        private final long deleteDate;
+
+        private Tombstone(final Index index, final long deleteDate) {
+            Objects.requireNonNull(index, "index must be set");
+            if (deleteDate < 0L) {
+                throw new IllegalArgumentException("invalid deleteDate [" + deleteDate + "]");
+            }
+            this.index = index;
+            this.deleteDate = deleteDate;
+        }
+
+        // create from stream
+        private Tombstone(StreamInput in) throws IOException {
+            index = new Index(in);
+            deleteDate = in.readLong();
+        }
+
+        /**
+         * The deleted index.
+         */
+        public Index getIndex() {
+            return index;
+        }
+
+        /**
+         * The date in milliseconds that the index deletion event occurred, used for logging/debugging.
+         */
+        public long getDeleteDate() {
```

Comment:
can we make the units clear in the name? (it's also not used now) getDeleteTimeInMillis (we typically use time)